File: 0502-ipo/0502-ipo.java
class Solution {
    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        int n = profits.length;
        ArrayList<int[]> arr = new ArrayList<>();
        
        
        for(int i=0;i<n;i++){
            arr.add(new int[]{capital[i],profits[i]});
        }
        
        Collections.sort(arr, Comparator.comparingInt(a -> a[0]));
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        
        int i = 0;

        while (k-- > 0) {
        
            while (i < n && arr.get(i)[0] <= w) {
                pq.add(arr.get(i)[1]);
                i++;
            }
        
            if (pq.isEmpty()) {
                break;
            }

            w += pq.poll();
        }

        return w;
    }
}

File: 2025-and-before/0502-ipo/0502-ipo.cpp
class Solution {
public:
    int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
        
        vector<array<int,2>>arr;
        for(int i=0;i<capital.size();i++){
            arr.push_back({capital[i],profits[i]});
        }
        
        sort(arr.begin(),arr.end());
        
        int i=0;
        priority_queue<int>pq;
        
        while(k--){
            
            while(i<arr.size() and arr[i][0]<=w){
                pq.push(arr[i][1]);
                i++;
            }
            
            if(pq.size() == 0)
                break;
            
            w+=pq.top();
            pq.pop();
        }
        
        return w;
    }
};